edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/DeviceScopeController.cs

```diff
@@ -0,0 +1,191 @@
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.Azure.Amqp.Framing;
+    using Microsoft.Azure.Devices.Edge.Hub.CloudProxy;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+    using Org.BouncyCastle.Security;
+
+    public class DeviceScopeController : Controller
+    {
+        static readonly string SupportedContentType = "application/json; charset=utf-8";
+
+        readonly Task<IEdgeHub> edgeHubGetter;
+
+        public DeviceScopeController(Task<IEdgeHub> edgeHub)
+        {
+            this.edgeHubGetter = Preconditions.CheckNotNull(edgeHub, nameof(edgeHub));
+        }
+
+        [HttpPost]
+        [Route("devices/{actorDeviceId}/modules/{actorModuleId}/devicesAndModulesInTargetDeviceScope")]
+        public async Task InvokeAsync([FromRoute] string actorDeviceId, [FromRoute] string actorModuleId, [FromBody] NestedScopeRequest request)
+        {
+            actorDeviceId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(actorDeviceId, nameof(actorDeviceId)));
+            actorModuleId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(actorModuleId, nameof(actorModuleId)));
+            Preconditions.CheckArgument(actorModuleId == Constants.EdgeHubModuleId);
+
+            EdgeHubScopeResult result = await this.HandleDevicesAndModulesInTargetDeviceScope(actorDeviceId, actorModuleId, request);
+            await this.SendResponse(result);
+        }
+
+        public async Task<EdgeHubScopeResult> HandleDevicesAndModulesInTargetDeviceScope(string actorDeviceId, string actorModuleId, NestedScopeRequest request)
+        {
+            Events.ReceivedScopeRequest(actorDeviceId, request);
+
+            EdgeHubScopeResult result = new EdgeHubScopeResult();
+
+            // Parse the target device ID from the authchain
+            if (!ValidateChainAndGetTargetDeviceId(actorDeviceId, request.AuthChain, out string targetDeviceId))
+            {
+                Events.InvalidAuthchain(request.AuthChain);
+                result.Status = HttpStatusCode.BadRequest;
+                return result;
+            }
+
+            // Actor device is claiming to be our child, and that the target device is its child.
+            // So we should have an authchain already cached for the target device.
+            IEdgeHub edgeHub = await this.edgeHubGetter;
+            Option<string> targetAuthChainOption = await edgeHub.GetAuthChainForIdentity(targetDeviceId);
+            string targetAuthChain = targetAuthChainOption.Expect(() => new UnauthorizedAccessException($"{targetDeviceId} is not a child of this Edge device"));
+
+            // The actor device should be somewhere in the target device's authchain
+            if (!targetAuthChain.Contains(actorDeviceId))
```

Comment:
What happens if target auth chain contains a partial match of actor device id, e.g. auth chain is ABC;DEF;GHI and actor device id is DE.  I think we should append delimiter at the beginning and end of auth chain and use contain with ";{actor device id};" value.

Reply:
good point, I think it's probably easier to read/maintain if I just tokenize the string and do a comparison against each element.

---
In reply to: [451899836](https://github.com/Azure/iotedge/pull/3149#discussion_r451899836) [](ancestors = 451899836)